### PR TITLE
FOUNDATIONS: Retreive All Teams

### LINKED ISSUE #87: FOUNDATIONS: Retreive All Team

### PR DESCRIPTION
**Pull request by GulchexrasGithub**
Closes #87

### CODE REVIEW

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+using System.Linq;
+using Tarteeb.Api.Models.Teams;
+using Tarteeb.Api.Models.Teams.Exceptions;
+using Xeptions;
+
+namespace Tarteeb.Api.Services.Foundations.Teams
+{
+    public partial class TeamService
+    {
+        private delegate IQueryable<Team> ReturningTeamsFunction();
```
Comment (Javlonbek-dev) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs:12]: Add Line

File: Tarteeb.Api/Models/Users/Exceptions/UserServiceException.cs [hunk 2]
```diff
@@ -3,15 +3,14 @@
...
 using Xeptions;
 
 namespace Tarteeb.Api.Models.Users.Exceptions
 {
     public class UserServiceException : Xeption
     {
-        public UserServiceException(Exception innerException)
-            : base(message: "User service error occured, contact support.", innerException)
+        public UserServiceException(Xeption innerException)
+        : base(message: "User service error occurred, contact support.", innerException)
```
Comment (jamshidbek001) [inline Tarteeb.Api/Models/Users/Exceptions/UserServiceException.cs:13]: Need Tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        private static string GetRandomString() =>
+            new MnemonicString().GetValue();
+
+        private static int GetRandomNumber() =>
+           new IntRange(min: 2, max: 10).GetValue();
+
+        private static DateTimeOffset GetRandomDateTimeOffset() =>
+            new DateTimeRange(earliestDate:
+                DateTime.UnixEpoch).GetValue();
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs:58]: 2 line

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        private IQueryable<Team> TryCatch(ReturningTeamsFunction returningTeamsFunction)
```
Comment (jamshidbek001) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs:13]: Need empty line

**Review by jasurbekyusuf — APPROVED** [review]

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public void ShouldThrowServiceExceptionOnRetrieveAllWhenAllServiceErrorOccursAndLogIt()
+        {
+            //given
+            string exceptionMessage = GetRandomString();
+
```
Comment (Javlonbek-dev) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs:54]: Remove Line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+               broker.SelectAllTeams()).Throws(sqlException);
+
+            //when
+            Action retrieveAllTeamsAction = () =>
+                this.teamService.RetrieveAllTeams();
+
+            //then
+            Assert.Throws<TeamDependencyException>(retrieveAllTeamsAction);
+
+            this.storageBrokerMock.Verify(broker => broker.SelectAllTeams());
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs:38]: Times.Once qolib ketibti

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+            this.loggingBrokerMock.Verify(broker =>
+                 broker.LogCritical(It.Is(SameExceptionAs(
+                    expectedTeamDependencyException))), Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs:45]: Bu pastda bo'ladi

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+            this.storageBrokerMock.Setup(broker =>
+               broker.SelectAllTeams()).Throws(sqlException);
+
+            //when
+            Action retrieveAllTeamsAction = () =>
+                this.teamService.RetrieveAllTeams();
+
+            //then
+            Assert.Throws<TeamDependencyException>(retrieveAllTeamsAction);
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs:36]: Bu eski standard. Yangisiga ko'chiring

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectAllTeams(), Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedteamServiceException))), Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs:81]: Fix order here

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.RetrieveAll.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+{
+    public partial class TeamServiceTests
+    {
+        [Fact]
+        public void ShouldRetrieveAllTeams()
+        {
+            //given
+            IQueryable<Team> randomTeams = CreateRandomTeam();
+            IQueryable<Team> storageTeams = randomTeams;
+            IQueryable<Team> expectedTeam = storageTeams;
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.RetrieveAll.cs:22]: expectedTeams

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.RetrieveAll.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.SelectAllTeams()).Returns(storageTeams);
+
+            //when
+            IQueryable<Team> actualTeam = this.teamService.RetrieveAllTeams();
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.RetrieveAll.cs:28]: actualTeams

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
+﻿//=================================
+// Copyright (c) Coalition of Good-Hearted Engineers
+// Free to use to bring order in your workplace
+//=================================
+
+using Microsoft.Data.SqlClient;
+using Moq;
+using System;
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs:8]: Order these

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        private static int GetRandomNumber() =>new IntRange(min: 2, max: 10).GetValue();
+
+        private static DateTimeOffset GetRandomDateTimeOffset() =>
+            new DateTimeRange(earliestDate: DateTime.UnixEpoch).GetValue();
+
+        private static Filler<Team> CreateTeamFiller()
+        {
+            var filler = new Filler<Team>();
+
+            filler.Setup().OnType<DateTimeOffset>().Use(GetRandomDateTimeOffset());
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs:61]: .Use(GetRandomDateTimeOffset()); put this code to next line or remove above Empty line

**Review by StudentofTATU — CHANGES_REQUESTED** [review]

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs (hunk 8 above)
Comment (StudentofTATU) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.RetrieveAll.cs]: Use      Should().BeEquivalentTo

https://github.com/ElbekDeveloper/Tarteeb/blob/master/Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Exceptions.RetrieveAll.cs

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.RetrieveAll.cs [hunk 14]
```diff
@@ -26,7 +26,7 @@ public void ShouldThrowCriticalDependencyExceptionOnRetrieveAllIfSqlErrorOccursA
                 new UserDependencyException(failedUserStorageException);
 
             this.storageBrokerMock.Setup(broker =>
-               broker.SelectAllUsers())
+                broker.SelectAllUsers())
                    .Throws(sqlException);
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.RetrieveAll.cs:30]: TAB

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.RetrieveAll.cs [hunk 15]
```diff
@@ -3,9 +3,9 @@
 // Free to use to bring order in your workplace
 //=================================
 
+using System.Linq;
 using FluentAssertions;
 using Moq;
-using System.Linq;
 using Tarteeb.Api.Models;
 using Xunit;
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.RetrieveAll.cs:11]: 22-line expectedUser  change to expectedUsers

Comment (StudentofTATU) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Logic.RetrieveAll.cs:11]: 22-line expectedUser  change to expectedUsers

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.RetrieveAll.cs (hunk 14 above)
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Users/UserServiceTests.Exceptions.RetrieveAll.cs:30]: TAB

**Review by StudentofTATU — APPROVED** [review]

**Pull request merged**